weave-js/src/components/Panel2/PanelTable/util.ts

```diff
@@ -60,29 +62,60 @@ export const stripTag = (type: Type): Type => {
   return isTaggedValue(type) ? taggedValueValueType(type) : type;
 };
 
-// Very simple type shape comparison
+// Very simple type shape comparison util used to determine if tableState
+// needs to be reset or not.
 export const typeShapesMatch = (type: Type, toType: Type): boolean => {
+  function propertyTypesAreCompatible(
+    incomingTypedDict: TypedDictType,
+    toTypedDict: TypedDictType
+  ) {
+    for (const key of Object.keys(toTypedDict.propertyTypes)) {
+      const keyType = incomingTypedDict.propertyTypes[key];
+      const toKeyType = toTypedDict.propertyTypes[key];
+      if (keyType === undefined || !typeShapesMatch(keyType, toKeyType!)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
   type = stripTag(type);
   toType = stripTag(toType);
+
+  // List type handling
   if (isList(type) || isList(toType)) {
-    if (!isList(type) || !isList(toType)) {
-      return false;
-    } else {
-      return typeShapesMatch(listObjectType(type), listObjectType(toType));
+    return (
+      isList(type) &&
+      isList(toType) &&
+      typeShapesMatch(listObjectType(type), listObjectType(toType))
+    );
+  }
+  // TypedDict type handling
+  if (isTypedDict(type) || isTypedDict(toType)) {
+    // Handle union<typedDict> case
+    //
+    // This is meant for special cases where a user adds another column to their table
+    // in a subsequent run.
+    if (isTypedDict(toType) && isUnion(type)) {
+      return type.members.every(
+        member =>
+          isTypedDict(member) &&
+          isTypedDict(toType) &&
+          propertyTypesAreCompatible(member, toType)
+      );
     }
-  } else if (isTypedDict(type) || isTypedDict(toType)) {
+
     if (!isTypedDict(type) || !isTypedDict(toType)) {
       return false;
-    } else {
-      for (const key of Object.keys(toType.propertyTypes)) {
-        const toKeyType = toType.propertyTypes[key]!;
-        const keyType = type.propertyTypes[key];
-        if (keyType === undefined || !typeShapesMatch(keyType, toKeyType)) {
-          return false;
-        }
-      }
     }
+
+    return propertyTypesAreCompatible(type, toType);
   }
+
+  // 2024/11/20, Note from Dom:
```

Comment:
this comment could be made more actionable by describing what case we return true here